pkg/gadgets/top/ebpf/tracer/tracer.go

```diff
@@ -0,0 +1,288 @@
+		t.iter.Close()
+	}
+
+	bpfstats.DisableBPFStats()
+}
+
+func getPidMapFromPids(pids []*piditer.PidIterEntry) map[uint32][]*types.PidInfo {
+	pidmap := make(map[uint32][]*types.PidInfo)
+	for _, e := range pids {
+		if _, ok := pidmap[e.ProgID]; !ok {
+			pidmap[e.ProgID] = make([]*types.PidInfo, 0, 1)
+		}
+		pidmap[e.ProgID] = append(pidmap[e.ProgID], &types.PidInfo{
+			Pid:  e.Pid,
+			Comm: e.Comm,
+		})
+	}
+	return pidmap
+}
+
+func getProgIDFromFile(fn string) (uint32, error) {
+	f, err := os.Open(fn)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+
+	sc := bufio.NewScanner(f)
+	for sc.Scan() {
+		if strings.HasPrefix(sc.Text(), "prog_id:") {
+			progID, _ := strconv.ParseUint(strings.TrimSpace(strings.Split(sc.Text(), ":")[1]), 10, 32)
+			return uint32(progID), nil
+		}
+	}
+	return 0, os.ErrNotExist
+}
+
+func getPidMapFromProcFs() (map[uint32][]*types.PidInfo, error) {
+	processes, err := os.ReadDir("/proc/")
+	if err != nil {
+		return nil, err
+	}
+	pidmap := make(map[uint32][]*types.PidInfo)
+	for _, p := range processes {
+		if !p.IsDir() {
+			continue
+		}
+		_, err := strconv.Atoi(p.Name())
+		if err != nil {
+			continue
+		}
+		fdescs, err := os.ReadDir(filepath.Join("/proc", p.Name(), "fdinfo"))
+		if err != nil {
+			continue
+		}
+		for _, fd := range fdescs {
+			if progID, err := getProgIDFromFile(filepath.Join("/proc", p.Name(), "fdinfo", fd.Name())); err == nil {
+				pid, _ := strconv.ParseUint(p.Name(), 10, 32)
+				if _, ok := pidmap[progID]; !ok {
+					pidmap[progID] = make([]*types.PidInfo, 0, 1)
+				}
+				comm, _ := os.ReadFile(filepath.Join("/proc", p.Name(), "comm"))
+				pidmap[progID] = append(pidmap[progID], &types.PidInfo{
+					Pid:  uint32(pid),
+					Comm: strings.TrimSpace(string(comm)),
+				})
+			}
+		}
+	}
+	return pidmap, nil
+}
+
+func (t *Tracer) nextStats() ([]types.Stats, error) {
+	stats := make([]types.Stats, 0)
+
+	var err error
+	var prog *ebpf.Program
+	var pids []*piditer.PidIterEntry
+	curID := ebpf.ProgramID(0)
+	nextID := ebpf.ProgramID(0)
+
+	curRuntimes := make(map[uint32]int64)
+	curRunCounts := make(map[uint32]uint64)
+
+	for {
+		nextID, err = ebpf.ProgramGetNextID(curID)
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				break
+			}
+			return nil, fmt.Errorf("could not get next program ID: %w", err)
+		}
+		if nextID <= curID {
+			break
+		}
+		curID = nextID
+		prog, err = ebpf.NewProgramFromID(curID)
+		if err != nil {
+			continue
+		}
+		pi, err := prog.Info()
+		if err != nil {
+			prog.Close()
+			continue
+		}
+
+		totalRuntime, _ := pi.Runtime()
+		totalRunCount, _ := pi.RunCount()
+
+		curRuntime := int64(0)
+		curRunCount := uint64(0)
+
+		// calculate delta, if possible
+		if oldrt, ok := t.prevRuntimes[uint32(curID)]; ok {
+			curRuntime = int64(totalRuntime) - oldrt
+		}
+		if oldctr, ok := t.prevRunCounts[uint32(curID)]; ok {
+			curRunCount = totalRunCount - oldctr
+		}
+
+		curRuntimes[uint32(curID)] = int64(totalRuntime)
```

Comment:
I'm a bit concerned that the ID of a program can be reused and it leads to a wrong result. What about using the tag to be sure the statistics are from the same program? (It's already a string, so concatenating it somehow with the programID should be straightforward.

Reply:
The previous runtimes/run counts are only valid for the next call of nextStats(), because I had the same concern. But I'll add the tag as well.